### PR TITLE
Brand new elevatin profile support for QField's measurement tool

### PR DESCRIPTION
**Pull request by nirvn**
This PR adds basic elevation profiling functionality to QField's measurement tool (a screencast of the UI/UX in action is available here: https://twitter.com/QFieldForQGIS/status/1584441812592119808).

Once this foundation is laid, a bunch of interesting features could be added. To name a few:
- tap on profile canvas to identify features
- tap along elevation profile to highlight location along the profiling curve on the map canvas
- overlay device/external GNSS location on the elevation profile
- draw elevation profile of a layer feature's line geometry from the feature form menu action
- etc.

### CODE REVIEW

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 5a6e71454e971ab7c0feb4822a6f09e3fee5109f: [**arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3501-fe66fd5b94bc07ff6995b1d6e8a159ac9c5df99d-arm64-android.apk)
<details>
<summary>Other architectures</summary>

- [arm-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3501-fe66fd5b94bc07ff6995b1d6e8a159ac9c5df99d-arm-android.apk)
- [x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3501-fe66fd5b94bc07ff6995b1d6e8a159ac9c5df99d-x64-android.apk)
- [x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3501-fe66fd5b94bc07ff6995b1d6e8a159ac9c5df99d-x86-android.apk)
</details>

**Comment by nirvn** [comment]
@m-kuhn , we got a build failure on windows due to some template declaration in a QGIS header:

```
C:\builddir\vcpkg_installed\x64-windows-static\include\qgis/qgstopologicalmesh.h(24,22): error : declaration does not declare anything [C:\builddir\src\core\qfield_core.vcxproj]
C:\builddir\vcpkg_installed\x64-windows-static\include\qgis/qgstopologicalmesh.h(25,22): error : declaration does not declare anything [C:\builddir\src\core\qfield_core.vcxproj]
C:\builddir\vcpkg_installed\x64-windows-static\include\qgis/qgstopologicalmesh.h(26,22): error : declaration does not declare anything [C:\builddir\src\core\qfield_core.vcxproj]
```

The exact lines are (https://github.com/qgis/QGIS/blob/c6008b7ee19117e374dce3679434abc0b1016d54/src/core/mesh/qgstopologicalmesh.h#L24):

```
#if defined(_MSC_VER)
template CORE_EXPORT QVector<int> SIP_SKIP;
template CORE_EXPORT QList<int> SIP_SKIP;
template CORE_EXPORT QVector<QVector<int>> SIP_SKIP;
#endif
```

Any clue as to how we overcome these errors?

File: src/core/qgsquick/qgsquickelevationprofilecanvas.cpp [hunk 1]
```diff
@@ -0,0 +1,723 @@
...
+#include "qgsabstractprofilesource.h"
+#include "qgsexpressioncontextutils.h"
+#include "qgsmaplayerelevationproperties.h"
+#include "qgsmaplayerutils.h"
+#include "qgsplot.h"
+#include "qgsprofilerenderer.h"
+#include "qgsprofilerequest.h"
+#include "qgsprojectelevationproperties.h"
+#include "qgsquickelevationprofilecanvas.h"
+#include "qgsterrainprovider.h"
```
Comment (m-kuhn) [inline src/core/qgsquick/qgsquickelevationprofilecanvas.cpp:29]:
```suggestion
#include <QVector>
#include <QList>
#include "qgsterrainprovider.h"
```

just a quick try

File: src/core/qgsquick/qgsquickelevationprofilecanvas.cpp [hunk 2]
```diff
@@ -0,0 +1,723 @@
+/***************************************************************************
+                          QgsQuickElevationProfileCanvas.cpp
+                          -----------------
+    begin                : March 2022
+    copyright            : (C) 2022 by Nyall Dawson
+    email                : nyall dot dawson at gmail dot com
```
Comment (m-kuhn) [inline src/core/qgsquick/qgsquickelevationprofilecanvas.cpp:6]: Not sure the metadata in the header is correct

File: src/core/qgsquick/qgsquickelevationprofilecanvas.cpp [hunk 3]
```diff
@@ -26,14 +26,17 @@
...
+#include <QList>
 #include <QQuickWindow>
 #include <QSGSimpleRectNode>
 #include <QSGSimpleTextureNode>
 #include <QScreen>
 #include <QTimer>
+#include <QVector>
 
+// Workaround a MSC compilation error
+#include "qgsterrainprovider.h"
```
Comment (nirvn) [inline src/core/qgsquick/qgsquickelevationprofilecanvas.cpp:39]: @m-kuhn , not working :/

**Comment by m-kuhn** [comment]
Excellent, how can you choose the dem raster?

**Comment by m-kuhn** [comment]
Question meant to be answered in docs ;-)

**Comment by nirvn** [comment]
@m-kuhn , the answer: https://docs.qgis.org/testing/en/docs/user_manual/working_with_raster/raster_properties.html?highlight=elevation#elevation-properties

We can link to a bunch of pages from QGIS docs in our own.

**Review by m-kuhn — APPROVED** [review]

**Pull request merged**